### DISCUSSION TITLE
Cannot use web driver, device is unknown

### DISCUSSION THREAD

**Discussion by eqMFqfFd** [php-webdriver/php-webdriver#848] [Q&A] [answered]
### Bug description

I have done a fresh install of Selenium, installed the jar and run it. The server is up and running in port 4444, howeever when I create the driver it outputs the following error:

```
 Fatal error:  Uncaught Facebook\WebDriver\Exception\SessionNotCreatedException: Unable to cr
eate new service: ChromeDriverService
Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:25:53'
System info: host: 'ubuntu', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'amd64', os.version: '4
.15.0-128-generic', java.version: '1.8.0_275'
Driver info: driver.version: unknown in /home/bot/vendor/php-webdriver/webdriver/lib/Exception/W
ebDriverException.php:125
```

Following the example that is in the README. I'm quite impressed that a program (from Facebook?) Fails with its installation guide, so I'm suspecting something related to configuration. I've run the example code though and it shouldn't fail.

I appreciate your help.

### How could the issue be reproduced

1. Follow the installation instructions.
2. Run the code.
3. See the error happen.

```php
$serverUrl = "http://localhost:4444/wd/hub";
$desiredCapabilities = DesiredCapabilities::chrome();
$desiredCapabilities->setCapability("acceptSslCerts", false);
$driver = RemoteWebDriver::create($serverUrl, $desiredCapabilities);
```

I should be able to use the Chrome web driver.

* Php-webdriver version: fb70876
* PHP version: PHP 8
* How do you start the browser: CLI
* Selenium server version: 
* Operating system:  Ubuntu 18.04.3 
* Browser used + version: Chrome.
* Browser driver: chromedriver

**Comment by andrewnicols** [comment]
Hi,

Can you please confirm what version of Selenium, Chrome, and ChromeDriver you are using?

The following code works for me using Selenium 3.141.59 with Chrome 87, and ChromeDriver 87.0.4280.20.

```
<?php

namespace Facebook\WebDriver;

use Facebook\WebDriver\Remote\DesiredCapabilities;
use Facebook\WebDriver\Remote\RemoteWebDriver;

require_once('vendor/autoload.php');

$serverUrl = "http://localhost:4444/wd/hub";
$desiredCapabilities = DesiredCapabilities::chrome();
$desiredCapabilities->setCapability("acceptSslCerts", false);
$driver = RemoteWebDriver::create($serverUrl, $desiredCapabilities);
```

Note that whilst this project uses the `Facebook` namespace, it was abandoned by Facebook and has been taken over by community volunteers. It is no longer developed by Facebook.

**Reply by eqMFqfFd** [reply]
> Hi,
> 
> Can you please confirm what version of Selenium, Chrome, and ChromeDriver you are using?
> 
> The following code works for me using Selenium 3.141.59 with Chrome 87, and ChromeDriver 87.0.4280.20.
> 
> ```
> <?php
> 
> namespace Facebook\WebDriver;
> 
> use Facebook\WebDriver\Remote\DesiredCapabilities;
> use Facebook\WebDriver\Remote\RemoteWebDriver;
> 
> require_once('vendor/autoload.php');
> 
> $serverUrl = "http://localhost:4444/wd/hub";
> $desiredCapabilities = DesiredCapabilities::chrome();
> $desiredCapabilities->setCapability("acceptSslCerts", false);
> $driver = RemoteWebDriver::create($serverUrl, $desiredCapabilities);
> ```
> 
> Note that whilst this project uses the `Facebook` namespace, it was abandoned by Facebook and has been taken over by community volunteers. It is no longer developed by Facebook.

I've installed the latest Jar, run the jar with 'java -jar selenium.jar` that's absolutely all what I've done, and run the code I sent you, nothing else. I'm unaware of chrome driver version, I'm new to the library.

Thanks for the last.

**Reply by andrewnicols** [reply]
Ah, sounds like you need to grab Chromedriver then - that's why Selenium is unable to start the service.

Chromedriver is the binary which sites between Selenium and Chrome and translates the W3C/JsonWire commands into Chrome actions.

You need the correct version of `chromedriver` for the version of Chrome that you are using, which you can find at https://chromedriver.chromium.org/downloads.

Alternatively you may want to just try this wrapper that I wrote: https://github.com/andrewnicols/chromedriver-wrapper
Essentialyl it checks the current version of Chrome and uses the correct version of `chromedriver` for the current version of Chrome, downloading any missing version as required.

**Reply by eqMFqfFd** [reply]
> Ah, sounds like you need to grab Chromedriver then - that's why Selenium is unable to start the service.
> 
> Chromedriver is the binary which sites between Selenium and Chrome and translates the W3C/JsonWire commands into Chrome actions.
> 
> You need the correct version of `chromedriver` for the version of Chrome that you are using, which you can find at https://chromedriver.chromium.org/downloads.
> 
> Alternatively you may want to just try this wrapper that I wrote: https://github.com/andrewnicols/chromedriver-wrapper
> Essentialyl it checks the current version of Chrome and uses the correct version of `chromedriver` for the current version of Chrome, downloading any missing version as required.

Just installed the latest download. The Chromedriver server started perfectly, but the program still crashed with the same error.

Chromedriver:
```
Only local connections are allowed.
Starting ChromeDriver 88.0.4324.27 (6347fe8bf1e48bd0c54d07dc55ca011cf40861c9-refs/branch-heads/4324@{#450}) on port 9515
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDriver safe.
ChromeDriver was started successfully.

```

**Reply by andrewnicols** [reply]
Is chromedriver in your path?

You can either run chromedriver directly, where you set the webdriver URL to something like `http://localhost:9515/`, or you can have Selenium do the heavy lifting, in which case the URL will be something like `http://localhost:4444/wd/hub`.

If you're doing the latter then Selenium will receive the request for a `chrome` browser and it will try to find and then start chromedriver, then it will control Chrome. If you're going this way then chromedriver _must_ be in the $PATH.

**Comment by eqMFqfFd** [comment]
@andrewnicols Thanks for the advice, you should put that on the guide because it is not there.

I have installed the latest chromedriver for Linux x64 in the same folder of Selenium and of my program with webdriver. I started chromedriver, then restarted Selenium and re-run my program. The error is exactly the same, unknown version of webdriver.

**Reply by OndraM** [reply]
> Thanks for the advice, you should put that on the guide because it is not there.

Hi there, what exactly have you been missing in documentation? 

You can also have a look to a whole guide we have describing options how to install and start Chrome: https://github.com/php-webdriver/php-webdriver/wiki/Chrome#installation

**Comment by eqMFqfFd** [comment]
@OndraM Thanks.

What I meant is that all the instructions to install the driver are not in the `README.md` but in the wiki. It could be easier access to link to the wiki in the part where the driver is started, because otherwise everyone will have my same issue.

@andrewnicols I followed the steps you gave me. With ChromeDriver up and Selenium "up and running", the error is:

```
PHP Fatal error:  Uncaught Facebook\WebDriver\Exception\WebDriverException: The driver server ha
s died. in /home/bot/vendor/php-webdriver/webdriver/lib/Remote/Service/DriverCommandExecutor.php
:50
```

**Reply by OndraM** [reply]
> What I meant is that all the instructions to install the driver are not in the README.md but in the wiki.  It could be easier access to link to the wiki in the part where the driver is started

Reference to the wiki right is right the first line of https://github.com/php-webdriver/php-webdriver#a-chromedriver ... 

> Below you will find a simple example. Make sure to read our wiki for more information on Chrome/Chromedriver.

Is this what you meant?

**Comment by eqMFqfFd** [comment]
@OndraM @andrefortin More logs:
```
PHP Fatal error:  Uncaught Facebook\WebDriver\Exception\UnknownErrorException: unknown error: Chrome failed to star
t: exited abnormally.
  (unknown error: DevToolsActivePort file doesn't exist)
  (The process started from chrome location /usr/bin/chromium-browser is no longer running, so ChromeDriver is assu
ming that Chrome has crashed.)
```

I appreciate your help.

**Comment by andrewnicols** [comment]
Hi @kenygamer,

Can you please describe the steps you have taken?

All instructions are indeed in the readme, but it is up to you as to which driver that you use, how you configure it, and how it is instantiated.

I would suggest, as a debugging step, the following:

1. do not run selenium directly
2. run `chromedriver --port=4444 -v`
3. configure php-webdriver to use a webdriver path of `http://localhost:4444`

It also sounds like you may be runing the wrong version of chromedriver for the version of chrome, or you may be running chrome from a different path than the one that chromedriver is using (`/usr/bin/chromium-browser`). It's hard to know because I don't know your setup.

**Comment by eqMFqfFd** [comment]
@andrewnicols Hello my friend,

After doing the steps you gave me, I get this error:
```
PHP Fatal error:  Uncaught Facebook\WebDriver\Exception\WebDriverException: JSON decoding of remote response failed.
Error code: 4
The response: '<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <link rel="stylesheet" type="text/css" href="/assets/displayhelpservlet.css" media="all"/>
  <link href="/assets/favicon.ico" rel="icon" type="image/x-icon" />
  <script src="/assets/jquery-3.1.1.min.js" type="text/javascript"></script>
  <script src="/assets/displayhelpservlet.js" type="text/javascript"></script>
  <script type="text/javascript">
    var json = Object.freeze('{"consoleLink": "\u002fwd\u002fhub","type": "Standalone","class": "org.openqa.grid.web.servlet.DisplayHelpHandler$DisplayHelpServletConfig",
"version": "3.141.59"}');
  </script>
</head>
<body>
<div id="content">
  <div id="help-heading">
    <h1><span id="logo"></span></h1>
    <h2>Selenium <span class="se-type"></span>&nbsp;v.<span class="se-version"></span></h2>
  </div>
  <div id="content-body">
    <p>
      Whoops! The URL specified route in /home/bot/vendor/php-webdriver/webdriver/lib/Remote/HttpCommandExecutor.php on line 338
```

**Comment by eqMFqfFd** [comment]
@andrewnicols To guide you more on the steps I have followed, I have this version of chromium browser installed with apt-get:
```
chromium-browser is already the newest version (87.0.4280.66-0ubuntu0.18.04.1).
```

I have downloaded the alpha 7 version of Selenium. I get this bunch of logs when I try to start up:

```
16:26:20.057 INFO [LogManager$RootLogger.log] - Using the system default encoding
16:26:20.059 INFO [LoggingOptions.getTracer] - Using OpenTelemetry for tracing
16:26:20.059 INFO [LoggingOptions.createTracer] - Using OpenTelemetry for tracing
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#noProviders for further details.
SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions prior to 1.8.
SLF4J: Ignoring binding found at [jar:file:/home/selenium/selenium.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#ignoredBindings for an explanation.
16:26:20.575 INFO [LoggingOptions.getTracer] - Using OpenTelemetry for tracing
Exception in thread "main" java.lang.NoSuchMethodError: java.io.FileReader.<init>(Ljava/io/File;Ljava/nio/charset/C
harset;)V
        at org.openqa.selenium.net.LinuxEphemeralPortRangeDetector.getInstance(LinuxEphemeralPortRangeDetector.java
:36)
        at org.openqa.selenium.net.PortProber.<clinit>(PortProber.java:42)
        at org.openqa.selenium.remote.service.DriverService$Builder.build(DriverService.java:401)
        at org.openqa.selenium.chrome.ChromeDriverService.createDefaultService(ChromeDriverService.java:119)
        at org.openqa.selenium.chrome.ChromeDriverInfo.isAvailable(ChromeDriverInfo.java:57)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566)
        at org.openqa.selenium.grid.node.config.NodeOptions.discoverDrivers(NodeOptions.java:187)
        at org.openqa.selenium.grid.node.config.NodeOptions.getSessionFactories(NodeOptions.java:84)
        at org.openqa.selenium.grid.node.local.LocalNodeFactory.create(LocalNodeFactory.java:66)
        at org.openqa.selenium.grid.commands.Standalone.createHandlers(Standalone.java:195)
        at org.openqa.selenium.grid.TemplateGridServerCommand.asServer(TemplateGridServerCommand.java:41)
        at org.openqa.selenium.grid.commands.Standalone.execute(Standalone.java:206)
        at org.openqa.selenium.grid.TemplateGridCommand.lambda$configure$2(TemplateGridCommand.java:98)
        at org.openqa.selenium.grid.Main.launch(Main.java:154)
        at org.openqa.selenium.grid.Main.go(Main.java:88)
        at org.openqa.selenium.grid.Main.main(Main.java:54)
```
Note that I am asked to specify a parameter (I used standalone, which ends up as `java -jar selenium.jar standalone`) since doing `java -jar selenium.jar` didn't work.

I have downloaded the latest beta of chromiumdriver, which looks like startup correctly. The version number starts with 88.0.

And you have adviced me to use no URI however when I do I get a JSON exception syntax error. Now that selenium is down it just hangs on a blank page.

Aside from selenium, everything else starts up well?

**Comment by OndraM** [comment]
Hi, I'm a bit puzzled what is going on there :).

Could you please provide steps you do:
- exact commands you run and their output
- what versions of the components you use (php-webdriver, selenium, chrome, chromedriver)
- what is the PHP code you execute and how do you start it

Also make sure you follow the steps from [README](https://github.com/php-webdriver/php-webdriver#getting-started) and from [Chrome installation](https://github.com/php-webdriver/php-webdriver/wiki/Chrome#installation) - you may be missing something.

Especially ensure you connect to the proper port and address where you remote end is running, [see README](https://github.com/php-webdriver/php-webdriver#2-create-a-browser-session).

**Comment by eqMFqfFd** [comment]
## Code to reproduce

```
$serverUrl = "http://localhost:4444/session";
$desiredCapabilities = DesiredCapabilities::chrome();
$desiredCapabilities->setCapability("acceptSslCerts", false);
$options = new ChromeOptions();
$options->setBinary("/home/selenium/chromedriver");
$desiredCapabilities->setCapability(ChromeOptions::CAPABILITY_W3C, $options);
$driver = RemoteWebDriver::create($serverUrl, $desiredCapabilities);
$driver->get("https://www.example.com");
var_dump($driver->findElements(WebDriverBy::id("img")));
```

## Commands
Selenium: `/usr/bin/java -jar /home/selenium/selenium.jar standalone`
Chromedriver: `/home/selenium/chromedriver`
Chromium is bundled in the system.

## Versions
PHP: `7.3.25`
Chromedriver: `88`
Chromium: `87.0.4280.66`
php-webdriver: `^1.9`

## Output
### Selenium
```
6:12:15.452 INFO [LogManager$RootLogger.log] - Using the system default encoding
06:12:15.456 INFO [LoggingOptions.getTracer] - Using OpenTelemetry for tracing
06:12:15.456 INFO [LoggingOptions.createTracer] - Using OpenTelemetry for tracing
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#noProviders for further details.
SLF4J: Class path contains SLF4J bindings targeting slf4j-api versions prior to 1.8.
SLF4J: Ignoring binding found at [jar:file:/home/selenium/selenium.jar!/org/slf4j/impl/StaticLoggerBinder
.class]
SLF4J: See http://www.slf4j.org/codes.html#ignoredBindings for an explanation.
06:12:16.418 INFO [LoggingOptions.getTracer] - Using OpenTelemetry for tracing
Exception in thread "main" java.lang.NoSuchMethodError: java.io.FileReader.<init>(Ljava/io/File;Ljava/nio
/charset/Charset;)V
        at org.openqa.selenium.net.LinuxEphemeralPortRangeDetector.getInstance(LinuxEphemeralPortRangeDet
ector.java:36)
        at org.openqa.selenium.net.PortProber.<clinit>(PortProber.java:42)
        at org.openqa.selenium.remote.service.DriverService$Builder.build(DriverService.java:401)
        at org.openqa.selenium.chrome.ChromeDriverService.createDefaultService(ChromeDriverService.java:1
19)
        at org.openqa.selenium.chrome.ChromeDriverInfo.isAvailable(ChromeDriverInfo.java:57)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:566)
        at org.openqa.selenium.grid.node.config.NodeOptions.discoverDrivers(NodeOptions.java:187)
        at org.openqa.selenium.grid.node.config.NodeOptions.getSessionFactories(NodeOptions.java:84)
        at org.openqa.selenium.grid.node.local.LocalNodeFactory.create(LocalNodeFactory.java:66)
        at org.openqa.selenium.grid.commands.Standalone.createHandlers(Standalone.java:195)
        at org.openqa.selenium.grid.TemplateGridServerCommand.asServer(TemplateGridServerCommand.java:41)
        at org.openqa.selenium.grid.commands.Standalone.execute(Standalone.java:206)
        at org.openqa.selenium.grid.TemplateGridCommand.lambda$configure$2(TemplateGridCommand.java:98)
        at org.openqa.selenium.grid.Main.launch(Main.java:154)
        at org.openqa.selenium.grid.Main.go(Main.java:88)
        at org.openqa.selenium.grid.Main.main(Main.java:54)
```
### Chromedriver
```
Starting ChromeDriver 88.0.4324.27 (6347fe8bf1e48bd0c54d07dc55ca011cf40861c9-refs/branch-heads/4324@{#450
}) on port 9515
Only local connections are allowed.
Please see https://chromedriver.chromium.org/security-considerations for suggestions on keeping ChromeDri
ver safe.
ChromeDriver was started successfully.
```

**Comment by OndraM** [comment, marked as answer]
Ok, thanks. It seems you are mixing multiple setups together.

1) `$serverUrl = "http://localhost:4444/session";` - there is no such case when this is correct URL. As I wrote in the previous post:

> Especially ensure you connect to the proper port and address where you remote end is running, [see README](https://github.com/php-webdriver/php-webdriver#2-create-a-browser-session).

As you used non-existing URL for the selenium server, it obviously cannot ever work.

2) Your Chromium and chromedriver version must match. If you have Chromium 87, you also must have chromedriver 87. This is stated [in our wiki](https://github.com/php-webdriver/php-webdriver/wiki/Chrome#installation), as well. Get the proper chromedriver [from here](https://chromedriver.chromium.org/downloads).

3) You should not start `selenium.jar` and `chromedriver` at the same time. Moreover, for the most basic usecases, you don't need to start neither of them manually and let php-webdriver handle it. See https://github.com/php-webdriver/php-webdriver/wiki/Chrome#start-directly-using-chromedriver-class . Then modify your code like this:

```php
putenv('WEBDRIVER_CHROME_DRIVER=/home/selenium/chromedriver'); // or wherever you have chromedriver binary

$driver = ChromeDriver::start();
$driver->get("https://www.example.com");
var_dump($driver->findElements(WebDriverBy::id("img")));
```

DON'T start chromedriver or selenium and just run the PHP script.

---

And a side note: All of this, including this example, is written in the [documentation](https://github.com/php-webdriver/php-webdriver/wiki/Chrome#start-directly-using-chromedriver-class), though.

**Reply by modnar1226** [reply]
Thank you for this. I've been using facebook/webdriver for some time now and made the jump to change the packages.
I however did not see this addition to the documentation. 

Using 
`ChromeDriver::start($capabilities); ` 
Worked perfectly, where I had been using symphony process to run the chromedriver manually on script start.

Mucho Kudos!!!!!

**Comment by eqMFqfFd** [comment]
@OndraM Can you explain what you mean by specific usecases? I need Selenium fully working. If I should start it or not before, I would like to know!

Also, when I do, Selenium gives me an error, do I ignore it? And is the php-webdriver going to start the servers as appropriate?

**Comment by eqMFqfFd** [comment]
@OndraM

**Reply by OndraM** [reply]
The use-cases are explained here, please read it: https://github.com/php-webdriver/php-webdriver/wiki/Chrome#start-chromedriver

"Selenium" (respectively the browser control) is of course fully working in all cases.
